pkg/policy/cell/policy_importer.go

```diff
@@ -0,0 +1,401 @@
+
+import (
+	"context"
+	"log/slog"
+	"net/netip"
+	"time"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/cilium/stream"
+
+	"github.com/cilium/cilium/pkg/container/set"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	ipcachetypes "github.com/cilium/cilium/pkg/ipcache/types"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/metrics"
+	"github.com/cilium/cilium/pkg/monitor/agent"
+	monitorapi "github.com/cilium/cilium/pkg/monitor/api"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/policy"
+	policytypes "github.com/cilium/cilium/pkg/policy/types"
+	"github.com/cilium/cilium/pkg/source"
+)
+
+type PolicyImporter interface {
+	UpdatePolicy(*policytypes.PolicyUpdate)
+}
+
+type policyImporterParams struct {
+	cell.In
+
+	Log       *slog.Logger
+	Lifecycle cell.Lifecycle
+	JobGroup  job.Group
+
+	Repo            policy.PolicyRepository
+	EndpointManager endpointmanager.EndpointManager
+	IPCache         *ipcache.IPCache
+	MonitorAgent    agent.Agent
+}
+
+type policyImporter struct {
+	log          *slog.Logger
+	repo         policy.PolicyRepository
+	epm          epmanager
+	ipc          ipcacher
+	monitorAgent agent.Agent
+
+	// prefixesByResources is the list of prefixes
+	// that belong to each resource. This is tracked separately
+	// so we can allocate and release prefixes as policy changes.
+	prefixesByResource map[ipcachetypes.ResourceID][]netip.Prefix
+
+	q chan *policytypes.PolicyUpdate
+}
+
+type ipcacher interface {
+	UpsertMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	RemoveMetadataBatch(updates ...ipcache.MU) (revision uint64)
+	WaitForRevision(ctx context.Context, rev uint64) error
+}
+
+type epmanager interface {
+	UpdatePolicy(idsToRegen *set.Set[identity.NumericIdentity], fromRev, toRev uint64)
+}
+
+func newPolicyImporter(cfg policyImporterParams) PolicyImporter {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	i := &policyImporter{
+		log:          cfg.Log,
+		repo:         cfg.Repo,
+		epm:          cfg.EndpointManager,
+		ipc:          cfg.IPCache,
+		monitorAgent: cfg.MonitorAgent,
+
+		q: make(chan *policytypes.PolicyUpdate, option.Config.PolicyQueueSize),
+
+		prefixesByResource: map[ipcachetypes.ResourceID][]netip.Prefix{},
+	}
+
+	cfg.Lifecycle.Append(cell.Hook{
+		OnStart: func(_ cell.HookContext) error {
+			go i.run(ctx)
+			return nil
+		},
+		OnStop: func(_ cell.HookContext) error {
+			cancel()
+			return nil
+		},
+	})
+
+	return i
+}
+
+// ResourceIDAnonymous is the anonymous ipcache resource used as a placeholder
+// for policies that allocate CIDRs but do not have an owning resource.
+// (This is only used for policies created by the local API).
+const ResourceIDAnonymous = "policy/anonymous"
+
+func (i *policyImporter) UpdatePolicy(u *policytypes.PolicyUpdate) {
+	i.q <- u
```

Comment:
Hmm. When could that happen? Is there a convention for this?

Reply:
I don't think there is convention What I have in mind here is that Hive is stopped, but the spawned go routines by the the K8s watcher might still be processing some incoming events before it realizes it was stopped, and thus might still attempt to call `UpdatePolicy` even though its context has been cancelled. If that happens and `i.q.` is full, the K8s/file watcher go routine might be stuck forever after a Hive shutdown.

(generally speaking Hive stops dependants before dependencies, but the K8s/file watcher go routtines are stopped asynchronously, so it won't wait for them to complete before the `policyImporter` cell is stopped)

In a real agent, we only shut down once (if SIGTERM is received), so leaking go routines there is "fine", but we also have a few tests that spawn an Hive during unit tests where dangling go routines might be an issue.

Reply:
Right, I think that's why the previous version of the code (which I cribbed from [here](https://github.com/cilium/cilium/blob/main/pkg/loadbalancer/experimental/reflector.go#L140)) drains the channel on shutdown, to prevent any last events from being stuck. I could close and drain the queue if you like.

Reply:
The alternative is to have an atomic bool that is set in `OnStop` and skip enqueuing here. Might be simpler

Reply:
I could close the channel on shutdown, but that risks causing a panic in the running agent, which would be bad. I'll take the risk of a leaked goroutine in future tests. Current testcases don't actually emit updates while shutting down anyways.